store/sqlstore/team_store.go

```diff
@@ -1504,6 +1504,32 @@ func (s SqlTeamStore) GetUserTeamIds(userId string, allowFromCache bool) ([]stri
 	return teamIds, nil
 }
 
+// GetCommonTeamIDsForUsers returns the intersection of all the teams to which the specified
+// users belong.
+func (s SqlTeamStore) GetCommonTeamIDsForUsers(userIDs []string) ([]string, error) {
+	if len(userIDs) != 2 {
+		return nil, errors.New("GetCommonTeamIDsForUsers can only be called with two users")
+	}
+	var teamIDs []string
+	const query = `
+	SELECT A.TeamId FROM (
+		SELECT TeamId FROM TeamMembers
+		WHERE UserId = :UserId0 AND DeleteAt = 0
+	) AS A INNER JOIN (
+		SELECT TeamId FROM TeamMembers
+		WHERE UserId = :UserId1 AND DeleteAt = 0
+	) AS B ON A.TeamId = B.TeamId
+	INNER JOIN Teams ON A.TeamId = Teams.Id
+	WHERE Teams.DeleteAt = 0`
```

Comment:
We are trying to avoid these sorts of query writing. I believe server team had initiative to use Squirrel to write queries.  For example, see `GetChannels` in `channel_store.go`. 
```Go
	query := s.getQueryBuilder().
		Select("Channels.*").
		From("Channels, ChannelMembers").
		Where(
			sq.And{
				sq.Expr("Id = ChannelId"),
				sq.Eq{"UserId": userId},
				sq.Or{
					sq.Eq{"TeamId": teamId},
					sq.Eq{"TeamId": ""},
				},
			},
		).
		OrderBy("DisplayName")
```
Try using squirrel to write out the query.
I'm sure there is examples in team_store too you can follow.

Reply:
I don't see any examples where squirrel is being used for nested queries. e.g. `getSearchGroupChannelsQuery` in `channel_store.go`. Trying to use squirrel for a nested query would likely require several SQL builders with multiple steps of error checking, which I think would make the code harder to read.

In general, I agree that using a query builder like squirrel can make a SQL query easier to read, especially if the shape of the query depends on certain parameters, but if the query is a constant string, and is formatted nicely, then I think just using that constant string is the most readable solution.

Reply:
Ok so I spent some time and was trying to think of a way to avoid nested sql. idk about others but personally for me I find it hard to read and a little bit hard to follow. I came up with this SQL example to avoid having nested. Is this suitable and does it work? Then we can take this and convert to squirrel. I'll still be approving this as I'm out next week and don't want to block this PR but have a thought about it and see if you can come up with a way to avoid nesting so we can use Squirrel ORM.

```
SELECT t.teamid
FROM   teammembers t
       INNER JOIN teammembers t2
               ON t.teamid = t2.teamid
       INNER JOIN teams t3
               ON t.teamid = t3.id
WHERE  t.userid = 'swrotanigt8gpg5wautz1s8zfo'
       AND t2.userid = '1igeyqqpu38f5mkwkyc7xhxqxy'
       AND t.deleteat = 0
       AND t3.deleteat = 0 
```

Reply:
I'm just not a big fan of writing out raw SQL queries and I personally always try to avoid nested SQL too cause its just a bit hard to follow when you're trying to understand what the query is really doing. 
I'm 2.5-3/5 on trying to come up with a non nested query that can be converted to squirrel.
If you find it very hard or too time consuming, then I'm okay with the way it is. I don't want it to become a blocker.

Reply:
The recommendation is to use squirrel for situations where you are dynamically constructing the query depending on several conditions.

For a simple straight query, anything is fine, since squirrel isn't giving you any advantage there. And for complicated queries, you need to spend more time conjuring the right squirrel commands. Normally people prefer raw SQL in these situations.

Reply:
@hahmadia 
😮  I like your query a lot better. And if I'm interpreting the EXPLAIN output correctly, Postgres is smart enough to filter each `TeamMembers` table first before performing the JOIN. Cool!